### DISCUSSION TITLE
clarification on recursive application

### DISCUSSION THREAD

**Discussion by proppy** [cuelang/cue#510] [General]
Starting with https://cue-review.googlesource.com/c/cue/+/6522 the following is resulting rightfully in a structural cycle error:
```
#sum: {
  A=arg: [...int]
  if len(A) > 1 {
    result: A[0] + (#sum & {arg: A[1:]}).result
  }
  if len(A) == 1 {
    result: A[0]
  }
}

sum1: (#sum & {arg: [1,2,3,4]}).result
```

But this doesn't:
```
#box: {
  V=value: int
  sum: {
    A=arg: [...#box]
    if len(A) > 1 {
      result: V + (A[0].sum & {arg: A[1:]}).result
    }
    if len(A) == 1 {
      result: V + A[0].value
    }
  }
}

box1: [...#box] & [for a in [1,2,3,4] {value: a}]
sum1: (box1[0].sum & {arg: box1[1:]}).result
```

And I'm not exactly clear if that's the expected behavior.

Is that fine to evaluate something recursively as long as we keep referring to different sub-struct unification, or should a bug should be filed against structural cycle detection.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/510.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
